ui/src/app/views/breadcrumb/component.ts

```diff
@@ -0,0 +1,121 @@
+import {Component} from '@angular/core';
+import {ActivatedRoute, NavigationEnd, PRIMARY_OUTLET, Router} from '@angular/router';
+
+export interface Breadcrumb {
+  label: string;
+  url: string;
+}
+
+@Component({
+  styleUrls: ['../../styles/buttons.css',
+    '../../styles/cards.css',
+    '../../styles/headers.css',
+    '../../styles/inputs.css',
+    './component.css'],
+  templateUrl: './component.html',
+  selector: 'app-breadcrumb'
+})
+
+export class BreadcrumbComponent {
+
+  breadcrumbs: Breadcrumb[];
+
+  constructor(
+      private activatedRoute: ActivatedRoute,
+      private router: Router) {
+    this.router.events.filter(event => event instanceof NavigationEnd).subscribe(event => {
+      this.breadcrumbs = this.buildBreadcrumbs(this.activatedRoute.root);
+    });
+  }
+
+  /**
+   * Generate a breadcrumb using the default label and url.
+   * Uses the route's paramMap to do any necessary variable replacement.
+   *
+   * @param {String} label
+   * @param {String} url
+   * @param {ActivatedRoute} route
+   * @returns {Breadcrumb}
+   */
+  private static makeBreadcrumb(
+      label: string,
+      url: string,
+      route: ActivatedRoute): Breadcrumb {
+    try {
+      const paramValues = route.paramMap['source']['_value'];
```

Comment:
This doesn't change `paramMap`, it only reads it. In the case where there are path variables in the url, those variables are defined in `route.paramMap['source']['_value']`. When the url value has path variables in it, this lets us pull them out of `paramMap` and put them into the breadcrumb label. This could definitely use some more explanatory comments. 

It might be simpler to call `paramMap.get('source')` - I'll look into making the lookup cleaner.